File: apps/android/app/src/main/java/pro/wewed/app/services/WeddingDayManifestTrustStore.kt
package pro.wewed.app.services

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

data class WeddingDayManifestKey(
    val keyId: String,
    val algorithm: String,
    val publicKeyDerBase64: String,
    val status: String,
    val activeFrom: String,
    val expiresAt: String? = null,
    val revokedAt: String? = null
)

data class VerifiedWeddingDayManifestTrust(
    val weddingId: String,
    val weddingShortId: String,
    val eventKey: String,
    val generatedAt: String,
    val expiresAt: String,
    val rootKeyId: String,
    val keys: List<WeddingDayManifestKey>
)

/**
 * LQR01 — observability of the cached offline authority a Gate device is admitting against.
 *
 * Offline admission is checked against the last root-verified manifest this device downloaded, so a
 * revocation made after [generatedAt] is invisible until the next refresh. This makes that age
 * visible to the usher. It is deliberately observability only: [isStale] never blocks admission and
 * the manifest's own expiry (enforced by [WeddingDaySyncService.verifyOfflinePass]) is unchanged.
 */
data class WeddingDayAuthorityStatus(
    val generatedAt: Date,
    val expiresAt: Date,
    val ageSeconds: Long,
    val isStale: Boolean,
    val isExpired: Boolean,
    val relyingOnCachedAuthority: Boolean = true
) {
    /** e.g. "Offline authority · generated 09:00 · expires 21:00 · age 3h (stale)". */
    fun summaryLine(timeZone: TimeZone = TimeZone.getDefault()): String {
        val clock = SimpleDateFormat("HH:mm", Locale.US).apply { this.timeZone = timeZone }
        val age = if (ageSeconds >= 3_600L) "${ageSeconds / 3_600L}h" else "${ageSeconds / 60L}m"
        val qualifier = when {
            isExpired -> " (expired)"
            isStale -> " (stale)"
            else -> ""
        }
        return "Offline authority · generated ${clock.format(generatedAt)} · " +
            "expires ${clock.format(expiresAt)} · age $age$qualifier"
    }

    companion object {
        /** Age after which cached authority is flagged stale. Observability only — not a TTL. */
        const val STALE_AFTER_SECONDS: Long = 2L * 60L * 60L

        const val CACHED_AUTHORITY_NOTICE =
            "Admission is being checked against this device's cached list; revocations made after " +
                "it was generated are not visible until refresh."

        fun from(trust: VerifiedWeddingDayManifestTrust, now: Long): WeddingDayAuthorityStatus? {
            val generated = parseWeddingDayIsoDate(trust.generatedAt) ?: return null
            val expires = parseWeddingDayIsoDate(trust.expiresAt) ?: return null
            val age = maxOf(0L, (now - generated.time) / 1_000L)
            return WeddingDayAuthorityStatus(
                generatedAt = generated,
                expiresAt = expires,
                ageSeconds = age,
                isStale = age >= STALE_AFTER_SECONDS,
                isExpired = expires.time <= now
            )
        }
    }
}

/**
 * Persistent cache containing only manifest metadata whose root signature has already been
 * verified. Pass-signing public keys therefore survive process death and support key rotation
 * without ever placing a private signing key on an usher device.
 */
class WeddingDayManifestTrustStore(
    private val storageDir: File? = null,
    private val gson: Gson = Gson()
) {
    private val mutex = Mutex()
    private val manifests = mutableMapOf<String, VerifiedWeddingDayManifestTrust>()
    private val file: File? = storageDir?.let { File(it, "wewed_wedding_day_trust.json") }

    init {
        load()
    }

    suspend fun save(manifest: VerifiedWeddingDayManifestTrust) = mutex.withLock {
        manifests[manifest.weddingId] = manifest
        persist()
    }

    suspend fun manifest(weddingId: String): VerifiedWeddingDayManifestTrust? = mutex.withLock {
        manifests[weddingId]
    }

    suspend fun signingKey(weddingId: String, keyId: String): WeddingDayManifestKey? = mutex.withLock {
        manifests[weddingId]?.keys?.firstOrNull { it.keyId == keyId }
    }

    suspend fun clear(weddingId: String) = mutex.withLock {
        manifests.remove(weddingId)
        persist()
    }

    private fun load() {
        val target = file ?: return
        if (!target.exists()) return
        try {
            val type = object : TypeToken<Map<String, VerifiedWeddingDayManifestTrust>>() {}.type
            val decoded: Map<String, VerifiedWeddingDayManifestTrust>? = gson.fromJson(target.readText(), type)
            manifests.clear()
            if (decoded != null) manifests.putAll(decoded)
        } catch (_: Exception) {
            manifests.clear()
        }
    }

    private fun persist() {
        val target = file ?: return
        storageDir?.mkdirs()
        target.writeText(gson.toJson(manifests))
    }
}

File: apps/ios/Wewed/Services/WeddingDayManifestTrustStore.swift
import Foundation

public struct WeddingDayManifestKey: Codable, Equatable, Sendable {
    public let keyId: String
    public let algorithm: String
    public let publicKeyDerBase64: String
    public let status: String
    public let activeFrom: String
    public let expiresAt: String?
    public let revokedAt: String?

    public init(
        keyId: String,
        algorithm: String,
        publicKeyDerBase64: String,
        status: String,
        activeFrom: String,
        expiresAt: String? = nil,
        revokedAt: String? = nil
    ) {
        self.keyId = keyId
        self.algorithm = algorithm
        self.publicKeyDerBase64 = publicKeyDerBase64
        self.status = status
        self.activeFrom = activeFrom
        self.expiresAt = expiresAt
        self.revokedAt = revokedAt
    }
}

public struct VerifiedWeddingDayManifestTrust: Codable, Equatable, Sendable {
    public let weddingId: String
    public let weddingShortId: String
    public let eventKey: String
    public let generatedAt: String
    public let expiresAt: String
    public let rootKeyId: String
    public let keys: [WeddingDayManifestKey]

    public init(
        weddingId: String,
        weddingShortId: String,
        eventKey: String,
        generatedAt: String,
        expiresAt: String,
        rootKeyId: String,
        keys: [WeddingDayManifestKey]
    ) {
        self.weddingId = weddingId
        self.weddingShortId = weddingShortId
        self.eventKey = eventKey
        self.generatedAt = generatedAt
        self.expiresAt = expiresAt
        self.rootKeyId = rootKeyId
        self.keys = keys
    }
}

/// How old the verified, cached manifest authority a gate is admitting against is.
///
/// Observability only. It never blocks admission and does not change the manifest's own expiry:
/// it exists so an operator can see that admission is being decided from this device's cached
/// list, and that revocations made after `generatedAt` are not visible until a refresh.
public struct WeddingDayAuthorityStatus: Equatable, Sendable {
    /// Age after which the cached authority is reported as stale.
    public static let staleAfter: TimeInterval = 2 * 60 * 60

    public static let cachedListNotice =
        "Admission is being checked against this device's cached list; revocations made after it was generated are not visible until refresh."

    public let generatedAt: Date
    public let expiresAt: Date
    /// Seconds since `generatedAt` (never negative).
    public let age: TimeInterval
    public let isStale: Bool
    public let isExpired: Bool
    /// Always true: offline admission is decided from the cached, root-signed manifest.
    public let relyingOnCachedAuthority = true

    /// Nil when the stored trust timestamps cannot be parsed — nothing is invented.
    public init?(
        trust: VerifiedWeddingDayManifestTrust,
        now: Date,
        staleAfter: TimeInterval = WeddingDayAuthorityStatus.staleAfter
    ) {
        guard let generatedAt = Self.parseIsoDate(trust.generatedAt),
              let expiresAt = Self.parseIsoDate(trust.expiresAt) else { return nil }
        self.generatedAt = generatedAt
        self.expiresAt = expiresAt
        self.age = max(0, now.timeIntervalSince(generatedAt))
        self.isStale = age >= staleAfter
        self.isExpired = now >= expiresAt
    }

    /// e.g. "Offline authority · generated 09:00 · expires 21:00 · age 3h (stale)".
    public func summaryLine(timeZone: TimeZone = .current) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = timeZone
        formatter.dateFormat = "HH:mm"
        let minutes = Int(age / 60)
        let ageText = minutes < 60 ? "\(minutes)m" : "\(minutes / 60)h"
        let qualifier = isExpired ? " (expired)" : (isStale ? " (stale)" : "")
        return "Offline authority · generated \(formatter.string(from: generatedAt))"
            + " · expires \(formatter.string(from: expiresAt))"
            + " · age \(ageText)\(qualifier)"
    }

    private static func parseIsoDate(_ iso8601: String) -> Date? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: iso8601) { return date }
        let fallback = ISO8601DateFormatter()
        fallback.formatOptions = [.withInternetDateTime]
        return fallback.date(from: iso8601)
    }
}

public actor WeddingDayManifestTrustStore {
    private var manifests: [String: VerifiedWeddingDayManifestTrust] = [:]
    private let fileURL: URL?

    public init(storageDirectory: URL? = nil) {
        if let storageDirectory {
            self.fileURL = storageDirectory.appendingPathComponent("wewed_wedding_day_trust.json")
        } else {
            self.fileURL = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)
                .first?
                .appendingPathComponent("wewed_wedding_day_trust.json")
        }

        if let fileURL,
           FileManager.default.fileExists(atPath: fileURL.path),
           let data = try? Data(contentsOf: fileURL),
           let decoded = try? JSONDecoder().decode([String: VerifiedWeddingDayManifestTrust].self, from: data) {
            manifests = decoded
        }
    }

    public func save(_ manifest: VerifiedWeddingDayManifestTrust) throws {
        manifests[manifest.weddingId] = manifest
        try persist()
    }

    public func manifest(weddingId: String) -> VerifiedWeddingDayManifestTrust? {
        manifests[weddingId]
    }

    public func signingKey(weddingId: String, keyId: String) -> WeddingDayManifestKey? {
        manifests[weddingId]?.keys.first(where: { $0.keyId == keyId })
    }

    public func clear(weddingId: String) throws {
        manifests.removeValue(forKey: weddingId)
        try persist()
    }

    private func persist() throws {
        guard let fileURL else { return }
        let directory = fileURL.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let data = try JSONEncoder().encode(manifests)
        try data.write(to: fileURL, options: .atomic)
    }
}
